### ISSUE TITLE
[Matmul][PCC] Matmul fails on PCC

### ISSUE THREAD

**Issue by mtopalovicTT** [tenstorrent/tt-metal#16366]
**Describe the bug**
We found during ResNet bringup on Forge.

When creating tensor using `torch.rand` I run into PCC error on below code:
```
import torch
import ttnn

from tests.ttnn.utils_for_testing import assert_with_pcc

device_id = 0
device = ttnn.open_device(device_id=device_id)

torch.manual_seed(0)
pcc = 0.99
torch_input_tensor_a = torch.rand([1, 2048])
torch_input_tensor_b = torch.rand([2048, 1000])
torch_output_tensor = torch.matmul(torch_input_tensor_a, torch_input_tensor_b)

input_tensor_a = ttnn.from_torch(torch_input_tensor_a, layout=ttnn.TILE_LAYOUT, device=device)
input_tensor_b = ttnn.from_torch(torch_input_tensor_b, layout=ttnn.TILE_LAYOUT, device=device)

output1 = ttnn.matmul(input_tensor_a, input_tensor_b)
output_tensor = ttnn.to_torch(output1)
assert_with_pcc(torch_output_tensor, output_tensor, pcc=pcc)

ttnn.close_device(device)
```

If I replace `torch.rand` with `torch.randn` I don't get PCC error.

**To Reproduce**

**Expected behavior**

**Screenshots**

**Please complete the following environment information:**
- OS: [e.g. Ubuntu 20.04]
- Version of software (eg. commit): I'm on 82b35a35473b1ce281fee4864c105648613043b7

**Additional context**

**Comment by mtopalovicTT** [comment]
@bbradelTT not sure if you are right person, but I see you are owner of a lot of matmul issues.
Do you have an idea what could be wrong with repro?

**Comment by bbradelTT** [comment]
@mtopalovicTT based on a quick test you need to increase the accuracy of the compute kernel.

Interestingly the `packer_l1_acc` being set reduces accuracy. @yugaoTT is this a known issue? If it is, any reasons to not change the default behaviour to have `packer_l1_acc` off?

The default is
```
WormholeComputeKernelConfig(math_fidelity=HiFi2,math_approx_mode=0,fp32_dest_acc_en=0,packer_l1_acc=1,dst_full_sync_en=0)
```

The smallest change would be to use the first compute kernel config in the test below.

```
def test_mm3(device):
    torch.manual_seed(0)
    pcc = 0.99
    torch_input_tensor_a = torch.rand([1, 2048])
    torch_input_tensor_b = torch.rand([2048, 1000])
    # The following works and would be the minimum required
    compute_kernel_config = ttnn.WormholeComputeKernelConfig(
            math_fidelity=ttnn.MathFidelity.HiFi2,
            math_approx_mode=False,
            fp32_dest_acc_en=False,
            packer_l1_acc=False,
        )
    # Most accurate
    compute_kernel_config = ttnn.WormholeComputeKernelConfig(
            math_fidelity=ttnn.MathFidelity.HiFi4,
            math_approx_mode=False,
            fp32_dest_acc_en=True,
            packer_l1_acc=False,
        )
    torch_output_tensor = torch.matmul(torch_input_tensor_a, torch_input_tensor_b)
            
    input_tensor_a = ttnn.from_torch(torch_input_tensor_a, layout=ttnn.TILE_LAYOUT, device=device)
    input_tensor_b = ttnn.from_torch(torch_input_tensor_b, layout=ttnn.TILE_LAYOUT, device=device)
            
    output1 = ttnn.matmul(input_tensor_a, input_tensor_b, compute_kernel_config=compute_kernel_config)
    print(f'{input_tensor_b}\n{output1}\n{torch_output_tensor}')
    output_tensor = ttnn.to_torch(output1)
    assert_with_pcc(torch_output_tensor, output_tensor, pcc=pcc)
```

**Comment by bbradelTT** [comment]
@mtopalovicTT yes, I'm the right person.

**Comment by mtopalovicTT** [comment]
Hi thanks for comment. I will wait for comment from @yugaoTT. This looks like something that needs to be on by default.

**Comment by yugaoTT** [comment]
it seems default DType for inputs are float32 when the tensor is created, this make sense now since you are accumulating bfp16 in L1 when packer_l1 enabled. without the flag, the accumulation is always done in DST in fp32 mode.

I think the question now for fp32 inputs, maybe we should auto-enable the fp32_dest_acc_en ?

**Comment by bbradelTT** [comment]
float32: enable fp32_dest_acc_en, disable packer_l1_acc
else: leave as is.

That will slow things down, but if you're using float32 you want the extra accuracy.

Makes sense.

**Comment by yugaoTT** [comment]
packer_l1_acc can still be on for float32 for better perf

q: should fp32_dest_acc_en be on when both inputs are float32, or when one of them is float32?

**Comment by bbradelTT** [comment]
Requirements:
1. when a compute kernel config is not provided, for float32 fp32_dest_acc_en should be on and packer_l1_acc should be off. Otherwise the current behaviour should be kept (fp32_dest_acc_en off and packer_l1_acc on).
2. Add a ttnn python unit test based on the description that fails before the change and works after the change.

**Comment by bbradelTT** [comment]
@mtopalovicTT @kmabeeTT we updated the default kernel config settings. Please check if this issue is fixed and let us know if it is not or close the issue.

**Comment by bbradelTT** [comment]
A quick update.

The two affected tests are
- tt-forge-fe fails with rtol fail (7.98% > 5%) in test test_mnist_training_with_grad_accumulation  ([link](https://github.com/tenstorrent/tt-forge-fe/actions/runs/12655690647/job/35266599313) need raw log)
- tt-torch fails with high atol (0.014 > 0.01) in a matmul test tests/torch/test_constant_fold.py::test_linear  ([link](https://github.com/tenstorrent/tt-torch/actions/runs/12655689098/job/35268675715#step:11:1731))

The tt-torch test requires HiFi4. We still need to find out if HiFi4 will work for the forge test.

Hopefully it'll work for them as well.

Then the next change needs to be:
ttnn/cpp/ttnn/operations/matmul/device/matmul_op.cpp in create_matmul_struct() change `math_fidelity` to be HiFi4 if increase_fidelity is true and both inputs are fp32

**Comment by bbradelTT** [comment]
@vsureshTT tt-forge-fe test also needs HiFi4. Please update the code to be HiFi4 if increase_fidelity is true and both inputs are fp32 and then run the pipelines to see that they pass.
